### PR TITLE
release worker first in standalone mode

### PR DESCRIPTION
**Pull request by jrobinAV**
This small change fixes the performance issue.

### CODE REVIEW

**Review by FredLL-Avaiga — DISMISSED** [review]
Well done!

**Review by FabienLelaquais — DISMISSED** [review]
👏👏👏

**Review by florian-vuillemot — DISMISSED** [review]

**Review by joaoandre-avaiga — DISMISSED** [review]

**Review by toan-quach — DISMISSED** [review]

**Review by gmarabout — DISMISSED** [review]
Well done!!!

**Comment by jrobinAV** [comment]
OK, after some more investigations, I better understand the issue.
The first fix (run release_worker callback before update_status) is not a correct solution; it is just a side effect that makes it work. The thing is that the real problem only appears when all the workers are occupied, and releasing the worker first hides the problem.

### Here is an explanation:

We have two callbacks to run after a job is completed.
1. `release_worker`: Releases the worker so it can be used to run another job.
2. `update_job_status`: Change the current job status to fail or completed, and update the subsequent job statuses to unblock them. This results in populating the queue of jobs to run with the next job.

Basically, the two callbacks must be completed to start the next job, as you can see in the following piece of code.

``` python
    def run(self):
        from time import sleep
        _TaipyLogger._get_logger().info("Start job dispatcher...")
        while not self._STOP_FLAG:
            try:
                if self._can_execute():
                    with self.lock:
                        job = self.orchestrator.jobs_to_run.get(block=True, timeout=0.1)
                    self._execute_job(job)
            except Exception:  # In case the last job of the queue has been removed.
                pass

    def _can_execute(self) -> bool:
        """Returns True if the dispatcher have resources to execute a new job."""
        return self._nb_available_workers > 0
```
The first callback, `release_worker,` makes the `_can_execute` method return `True` when the worker is released. The second callback, `update_job_status`, populates the `self.orchestrator.jobs_to_run`.

In the code above, when all the workers are occupied, the `if` condition returns `False` and directly retries without sleeping at all. This is bad practice and consumes a lot of CPU, so other threads are slower. This is what we observed when the `update_job_status` is executed first. The callback thread is slower. It is not the case when it is executed after the `release_worker`, because the previous code does not directly loop. It is sleeping thanks to the `self.orchestrator.jobs_to_run.get()` before looping again.

### The solution is easy:
We just need to sleep to go faster!

``` python
    def run(self):
        from time import sleep
        _TaipyLogger._get_logger().info("Start job dispatcher...")
        while not self._STOP_FLAG:
            try:
                if self._can_execute():
                    with self.lock:
                        job = self.orchestrator.jobs_to_run.get(block=True, timeout=0.1)
                    self._execute_job(job)
                else:
                    sleep(0.1)
            except Exception:  # In case the last job of the queue has been removed.
                pass

    def _can_execute(self) -> bool:
        """Returns True if the dispatcher have resources to execute a new job."""
        return self._nb_available_workers > 0
```

**Review by gmarabout — APPROVED** [review]
Awesome!! Well done @jrobinAV 👏🏻

**Comment by FredLL-Avaiga** [comment]
does this work with sleep(0) ?

File: taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py [hunk 1]
```diff
@@ -69,6 +70,8 @@ def run(self):
                             break
                         job = self.orchestrator.jobs_to_run.get(block=True, timeout=0.1)
                     self._execute_job(job)
+                else:
+                    time.sleep(0.1)  # We need to sleep to avoid busy waiting.
```
Comment (gmarabout) [inline taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py:74]: 💭 Maybe this value should be configurable via env variable/configuration.

**Review by gmarabout — APPROVED** [review]

File: taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py:74]: Not sure that would bring any value

**Review by FredLL-Avaiga — APPROVED** [review]

File: taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py (hunk 1 above)
Comment (gmarabout) [reply taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py:74]: I think it might. Here we arbitrarily wait for 0.1 second. Why not 0.01 ? Or 1 second?

Comment (gmarabout) [reply taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py:74]: The number of workers is configurable. I see no reason for wait time in dispatcher loop not to be configurable as well. 
It should just come with reasonable defaults (and `0.1` looks a reasonable value).

**Review by gmarabout — APPROVED** [review]

File: taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py:74]: I think this `sleep()` is really technical ie it allows to give back control to the system and the value is not really important here: it should be the minimum possible.
That's why I asked if sleep(0) was working ...

**Review by trgiangdo — APPROVED** [review]

File: taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py (hunk 1 above)
Comment (toan-quach) [reply taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py:74]: I agree with Gregoire, though then it makes the issue a bit more complicated as too much option isn't necessarily a good thing. If we decide to work on it, maybe a different PR would be better too

**Review by toan-quach — APPROVED** [review]

File: taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py (hunk 1 above)
Comment (gmarabout) [reply taipy/core/_orchestrator/_dispatcher/_job_dispatcher.py:74]: I totally agree ! This is not blocking for me and could eventually be addressed later if there is a need.

**Pull request merged**